Distributed-displaced mesh generation plus mortar convergence studies

lindsayad opened the pull request:
Looks at:
- approximately matching size discretizations on slave/master
- approximately 2:1 slave:master mismatch
- approximately 1:2 slave:master mismatch

I'm observing first order convergence for the LM no matter which I use,
when I expect super-convergence especially in the case of matching
mesh sizes.

![master-coarser](https://user-images.githubusercontent.com/10248304/82700714-9cff8000-9c23-11ea-92f6-bf6100fff4f9.png)
![slave-coarser](https://user-images.githubusercontent.com/10248304/82700715-9d981680-9c23-11ea-8c6a-0a0ff1a7068c.png)
![same](https://user-images.githubusercontent.com/10248304/82700717-9d981680-9c23-11ea-829b-8d2a2dcb168b.png)

I saw super-convergence in results shared in #13080 so I'm going to have to spend some more time with this

lindsayad committed:
Make mms more flexible

lindsayad committed:
Standard LM p1p0 convergence study

Looks at:
- approximately matching size discretizations on slave/master
- approximately 2:1 slave:master mismatch
- approximately 1:2 slave:master mismatch

I'm observing first order convergence for the LM no matter which I use,
when I expect super-convergence especially in the case of matching
mesh sizes.

Refs #13080

lindsayad commented:
This is if I have *exact* 2:1 mismatches. It seems like there is almost certainly a bug in here...
![master-coarser](https://user-images.githubusercontent.com/10248304/82701459-18156600-9c25-11ea-9c0a-4cef3314a2d9.png)
![slave-coarser](https://user-images.githubusercontent.com/10248304/82701462-18adfc80-9c25-11ea-9ceb-a16dbe7b7b92.png)

@dewenyushu just so you're aware that I'm doing some convergences studies with a standard LM. Some results I'm not pleased with...

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/532890/) on e2f586a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15338/site)

This comment will be updated on new commits.

lindsayad committed:
Add old gmsh mesh example used for results in #13080

This guy looks ok. Using p1p0, we have primal second order convergence
and 1.5 order convergence for lambda

lindsayad commented:
Ok the results from a38f842 look like what they did in #13080, so that's a relief. Maybe there's a bug in how I'm using the mesh generators in 59e67ab 

![p1p0](https://user-images.githubusercontent.com/10248304/82719300-1caa4080-9c5e-11ea-904b-a8dd82211922.png)

lindsayad committed:
Add p2p1 gmsh case

lindsayad commented:
3aff05f with p2p1
![p2p1](https://user-images.githubusercontent.com/10248304/82719475-c50cd480-9c5f-11ea-822b-bcf223ca40a2.png)

lindsayad commented:
@jwpeterson it looks like the mortar implementation isn't totally broken

jwpeterson commented:
@lindsayad I started with your case of a 5x11 mesh next to a 5x5 mesh (so approximately 2:1 mismatch) and did a uniform refinement study of the stabilized P1-P0 formulation, varying which side of the interface was treated as the slave side. What I found was that both formulations converged at O(h) in the asymptotic limit, so no super-convergence, but the error for the "coarser slave" case was *much* (like 100x) lower. I'd be curious to see if you get a similar result in the existing and dual formulations in MOOSE.

![compare_2to1](https://user-images.githubusercontent.com/1775907/82853425-5e8bee80-9ecb-11ea-9c98-aeb5b984e418.png)

lindsayad commented:
> so no super-convergence

Does the technical report outline conditions under which super-convergence may be observed?

> but the error for the "coarser slave" case was _much_ (like 100x) lower.

Very interesting. Here's a new result with my "gmsh" mesh case:
![p1p0-multiple-lines](https://user-images.githubusercontent.com/10248304/82929125-0a6b2380-9f38-11ea-922e-44c669fbf6e7.png)

I do indeed see that the coarser slave yields lower errors for the mesh sizes examined; however, in this case it appears that the asymptotic convergence rate is better for the coarse master case.

lindsayad commented:
> stabilized P1-P0 formulation

Is stabilization technically required even when using a mixed order discretization?

lindsayad commented:
Ok, yea with the "MOOSE" mesh, I see basically the exact same thing as you did @jwpeterson :
![multiple-lines](https://user-images.githubusercontent.com/10248304/82932192-e2ca8a00-9f3c-11ea-8a30-6a70b525d8df.png)

jwpeterson commented:
Do your Gmsh meshes not have the 2:1 element size mismatch or something? I'm curious why they are so "good" compared to the MOOSE mesh. Any other difference between the Gmsh and MOOSE meshes?

> Is stabilization technically required even when using a mixed order discretization?

The unequal-order discretizations are still technically LBB-unstable, as far as I understand things. There's more info on this in the Barbosa & Hughes paper cited in the report. That doesn't mean that you can't get an answer, though, so "required" is a strong word. 

I always observed much lower error for the stabilized formulations, e.g. by about a factor of 10 in the 2:1 mismatch case. You can see this especially in the line plots comparing the flux solutions in the report... also when using unstabilized P^1 - P^0, I sometimes observed a large increase in the number of GMRES iterations required by the linear solver. I wasn't sure exactly what caused this, but generally speaking the stabilized formulation gave much more predictable results so I always preferred to use it.

jwpeterson commented:
> Does the technical report outline conditions under which super-convergence may be observed?

No, the most useful a-priori error estimate that I found was in the Barbosa & Hughes paper, which is Eqn. (67) in the report. Technically it only applies to the stabilized formulation and the mesh-dependent norm shown, but it is qualitatively similar to other a-priori estimates in that only the "combined" primal H1-error + flux L2-error is predicted. So super-convergence of the flux error is never precluded by these estimates, but also they don't allow you to predict when it will/will not happen. 

So far in your results I have always seen at least O(h^2) for the flux in the P2-P1 formulation and at least O(h) for the flux in the P1-P0 formulation, which is in agreement with the theory. The theory also doesn't preclude (but doesn't explain) the difference we see in the coarse slave vs. coarse master formulations, this must somehow be accounted for by the constant "C" in the estimate being quite different for the two cases...

jwpeterson commented:
Anyway, I think we can conclude from your results that choosing the slave side to be the coarser of the two almost always results in lower overall error, except for possibly in the P2-P1 case on "sufficiently-refined" grids but in that case the error is very close for almost the entire range of h, so this may be something to worry about only in the P1-P0 case.

lindsayad commented:
> Do your Gmsh meshes not have the 2:1 element size mismatch or something? I'm curious why they are so "good" compared to the MOOSE mesh. Any other difference between the Gmsh and MOOSE meshes?

They do not have the 2:1 mismatch; it's probably more like 6:5, or maybe even closer than that to a unity ratio (but definitely not perfectly 1). In terms of element size mismatch the gmsh mesh is probably closest to the "same" case for the MOOSE mesh, where there is still a slight mismatch with the master begin slightly finer. I guess then that the MOOSE and gmsh mesh results are not too discrepant, since gmsh "slightly coarser slave" case has a flux converging at ~1.1 while the MOOSE "slightly coarser slave" (e.g. the poorly named "same") converges at 1.0

jwpeterson commented:
> They do not have the 2:1 mismatch; it's probably more like 6:5, or maybe even closer than that to a unity ratio (but definitely not perfectly 1).

OK, that's good, I think it explains why we see superconvergent flux behavior for the Gmsh meshes. This is also consistent with what I've seen in the past for nearly-equal-h-but-mismatched meshes. 

It would be great if you could also run the P2-P1 solve on the 2:1 mismatched "MOOSE" meshes and see if you observe the same big accuracy difference between the coarse and fine slave in that case. Based on your other results, I would expect the difference to be much less pronounced than in the P1-P0 case, but it would be good to confirm that. Unfortunately, I did not implement P2-P1 in my libmesh code, so I don't have a point of comparison for this one.

lindsayad commented:
This is P2P1 with the MOOSE generated mesh with slave-to-master size ratio 11:5
![multiple-lines](https://user-images.githubusercontent.com/10248304/82941529-596e8400-9f4b-11ea-9cce-c48793e4ec4f.png)

Wowzers trousers. You could say that the difference is still quite pronounced haha. Even worse here...

jwpeterson commented:
> You could say that the difference is still quite pronounced haha.

To me this looks like a definite bug in the "coarse slave" case since the rates are wrong... no idea what would cause that though given that coarse master seems to work fine!

lindsayad committed:
Add MOOSE-generated mesh case for P2P1

This shows zeroth order convergence for the LM variable when the
slave side is coarser. John does not see this with his libMesh code. He
sees what you should see (e.g. O(h) to O(h^1.5)). So there is definitely
a bug in the mortar code. These will serve as good test cases for
ascertaining when the bug is fixed

lindsayad committed:
Add stabilization

lindsayad commented:
Oh yea, stabilization is the key!!! Thanks @jwpeterson for getting me to see the light. Below is P1P0 exact 2:1 mismatch. Stabilized has delta = 0.4:
![4-to-2](https://user-images.githubusercontent.com/10248304/83201238-e9523080-a0f9-11ea-86da-2ab2c38340be.png)

lindsayad commented:
> To me this looks like a definite bug in the "coarse slave" case since the rates are wrong... no idea what would cause that though given that coarse master seems to work fine!

This definitely still seems to be the case for P2P1. Stabilization did not make the bad P2P1 coarse slave behavior go away. I actually didn't catch that it was the coarse slave case that was buggy in P2P1; I had been thinking that it was the coarse master case that was buggy as it was for P1P0 (whose "bug" was fixed by just adding stabilization (see https://github.com/idaholab/moose/pull/15338#issuecomment-635646775)).

Here's an updated P2P1 plot, with stabilization cases also included:
![multiple-lines](https://user-images.githubusercontent.com/10248304/83215943-f16f9780-a11c-11ea-88c1-64814bd8322b.png)

Although the error is still significantly higher than for a roughly matched interface, adding stabilization the the coarse master case drops the error substantially from the unstabilized case (and actually makes it slightly super-convergent with slope 2.3).

But stabilization had no effect on the coarse slave case for P2P1, the flux is still non-convergent and the primal variable has sub-optimal convergence.

lindsayad commented:
Ok heretofore summary. The below figures have a base mesh of left-nx=2, right-nx=2, coarse-ny=2. In the plots ny denotes the fine side ny:

![p2p1-ny-4](https://user-images.githubusercontent.com/10248304/83220364-9f803f00-a127-11ea-8e01-83e9528418a6.png)
![p2p1-ny-5](https://user-images.githubusercontent.com/10248304/83220365-a018d580-a127-11ea-94fd-032c7e18c7c7.png)
![p1p0-ny-4](https://user-images.githubusercontent.com/10248304/83220376-a5762000-a127-11ea-9d10-18b203c61c5c.png)
![p1p0-ny-5](https://user-images.githubusercontent.com/10248304/83220379-a60eb680-a127-11ea-8c74-f7e9a8918c7f.png)

P1P0 looks good. Don't quite know what to make of P2P1. The 5:2 mismatch looks pretty wonky but the 4:2 looks a lot better. Maybe for mortar validation I shouldn't be doing solution continuity at all, and maybe I could figure out an MMS for something like gap conductance...

lindsayad committed:
Fix mortar Jacobians

lindsayad committed:
set_title method to ConvergencePlot

lindsayad committed:
Still trying to verify...

jwpeterson commented:
@lindsayad I agree the P1-P0 results are a big improvement, nice work! I will check whether the Barbosa & Hughes paper says anything special about P2 elements, but I don't seem to recall anything. It is a bit surprising that P2-P1 converges for a little while but then the line turns back up... is it possible there is some tolerance issue in the nodal projections that is only showing up once the mesh is sufficiently refined?

lindsayad commented:
Ok got some weekend free time to return to this. I setup a gap conductance MMS case, which nicely doesn't require stabilization, and I'm seeing the same behavior that I did with the solution continuity test case. P1P0 looks great. P2P1 looks good for a 4:2 mismatch but like trash for 5:2. Something I find interesting here is that the coarse master case has lower error for the flux in all cases.

![p1p0-ny-4](https://user-images.githubusercontent.com/10248304/83960416-6fc6ea80-a83d-11ea-8ae9-7c7ed3952266.png)
![p1p0-ny-5](https://user-images.githubusercontent.com/10248304/83960417-70f81780-a83d-11ea-8a81-3fe158bddb97.png)
![p2p1-ny-4](https://user-images.githubusercontent.com/10248304/83960420-810ff700-a83d-11ea-8825-7ea29652ae51.png)
![p2p1-ny-5](https://user-images.githubusercontent.com/10248304/83960421-81a88d80-a83d-11ea-9b57-dbdece13d821.png)

lindsayad committed:
gap conductance mms study

lindsayad commented:
Let's see, the coarse slave cases work out like this:

- optimal convergence: 2:2, 3:2, 4:2
- poor convergence: 5:2, 6:2, 8:2

jwpeterson commented:
This sounds like good progress, especially the fact that we can rule out issues with stabilization since you used a gap heat conduction MMS. The fact that *all* coarse master cases and *some* coarse slave cases are working suggests to me that the code is essentially working, but there is some bug which seems to only be triggered by P2 elements with a coarser slave discretization. 

I'd suspect either the projection of slave nodes onto the master surface or the inverse projection of master nodes from the slave surface. If there is a bug there, the general formulas for the projections are most likely correct and there are just some "edge" cases which don't work correctly. What do you think? Any ideas what the issue might be?

jwpeterson commented:
OK, one more comment: after drawing some pictures, it looks like all of your sub-optimal cases occur when at least 3 master elements touch a *single* element on the slave side, the simplest example of this being the 3:1 case. In all the optimal cases a maximum of 2 master side elements touch any given slave element. Maybe this is a clue to what the issue is...

lindsayad commented:
Oh wow such beautiful code

![p2p1-7-to-3](https://user-images.githubusercontent.com/10248304/84068776-59747800-a97e-11ea-8c6a-b80be71a2986.png)
![p2p1-6-to-3](https://user-images.githubusercontent.com/10248304/84068774-58434b00-a97e-11ea-922a-35d0d1187cd8.png)
![p2p1-5-to-3](https://user-images.githubusercontent.com/10248304/84068778-59747800-a97e-11ea-8531-9d37c56a2676.png)
![p2p1-4-to-3](https://user-images.githubusercontent.com/10248304/84068780-5a0d0e80-a97e-11ea-9986-0bb477188dea.png)
![p2p1-3-to-3](https://user-images.githubusercontent.com/10248304/84068784-5a0d0e80-a97e-11ea-89af-8196feb54ed2.png)
![p2p1-2-to-3](https://user-images.githubusercontent.com/10248304/84068788-5aa5a500-a97e-11ea-8374-ffcf283d4b34.png)
![p2p1-1-to-3](https://user-images.githubusercontent.com/10248304/84068789-5b3e3b80-a97e-11ea-8bb3-e23645ad6c40.png)

lindsayad committed:
Fix bux in interior node placement for EDGE3s

jwpeterson commented:
Looks good, we expect O(h^2) convergence rate for the flux in this case and it looks like we get exactly that now! I'm glad you did the extensive MMS testing... otherwise this bug would probably have gone undetected for even longer than it did.

lindsayad commented:
Yea I'm glad @dewenyushu's work inspired me to do it. It's been fun to work on mortar again!

jwpeterson commented:
Maybe the only case you haven't explicitly confirmed is P2-P0, for which I would expect to see O(h^1.5) rate of convergence for the flux. 

The more I think about it, in the P2-P0 case you get the biggest bang for your buck since you get the full O(h^{r + 3/2}) (where r is the polynomial degree of the flux var) whereas with P2-P1 you get P1-P0 you get O(h^{r+1}) only. IMO that makes P2-P0 an interesting case.

lindsayad commented:
I'll try that out after I figure out a sparsity allocation problem that
makes this annoyingly slow to run for large problems...

On Mon, Jun 8, 2020 at 12:16 PM John W. Peterson <notifications@github.com>
wrote:

> Maybe the only case you haven't explicitly confirmed is P2-P0, for which I
> would expect to see O(h^1.5) rate of convergence for the flux.
>
> The more I think about it, in the P2-P0 case you get the biggest bang for
> your buck since you get the full O(h^{r + 3/2}) (where r is the polynomial
> degree of the flux var) whereas with P2-P1 you get P1-P0 you get O(h^{r+1})
> only. IMO that makes P2-P0 an interesting case.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/15338#issuecomment-640820011>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGA4ANIE6FGW44TXPARYLRVU2H7ANCNFSM4NICAFNQ>
> .
>

lindsayad commented:
It looks to me that with P2P0 you actually lose optimal convergence in the primal variable
![p2p0-2-to-2](https://user-images.githubusercontent.com/10248304/84089542-13caa600-a9a4-11ea-821a-939cfd10a69a.png)
![p2p0-3-to-2](https://user-images.githubusercontent.com/10248304/84089540-13320f80-a9a4-11ea-89e4-c86e8843c371.png)
![p2p0-4-to-2](https://user-images.githubusercontent.com/10248304/84089539-12997900-a9a4-11ea-9ee9-b688d7360806.png)
![p2p0-6-to-2](https://user-images.githubusercontent.com/10248304/84089534-0f05f200-a9a4-11ea-9fbd-1c6727e4e4e0.png)

Taking the 5:2 coarse master case through a couple more refinements:
![p2p0-5-to-2](https://user-images.githubusercontent.com/10248304/84089600-3b217300-a9a4-11ea-9503-acfe1eb918b9.png)

lindsayad committed:
Remove mallocs during MatSetValues for mortar

lindsayad commented:
Convergence summary:
| discretization | L2 primal | L2 lambda |
| --- | --- | --- |
| p2p2 | 3 | 3|
|p2p1 |3|2|
|p2p0|2|1|
|p1p1|2|2|
|p1p0|2|1|

lindsayad committed:
Automate convergence tests for gap-conductance

jwpeterson commented:
Hmm... I'm not quite sure what to make of the P2-P0 results. Not what I was expecting but it's definitely possible I (still) don't completely understand the a-priori estimates for this case. Note that  you are computing the L2-norm of the primal solution in all of these plots, but the estimates are actually for (something proportional to) the H1-norm of the primal solution, for what it's worth. 

1. For the stabilized solution continuity case, the relevant norm and estimate are given by (66), (67), respectively, in the Tech Report.
1. For the unstabilized equal-order case, the norm and estimate are given by (58), (65).

I think neither of these estimates are directly applicable to the P2-P0 gap heat conduction problem, unfortunately, since it uses non-equal order approximations *and* doesn't use stabilization.  

The latter estimate *should* be applicable to the P2-P2 discretization case, and it predicts that you should see O(h^2) convergence in the combined norm. Therefore I was a bit surprised that you saw O(h^3) for lambda in that case, but then I remembered that you are using that quartic exact solution? I believe the true flux is quadratic in that case, so it is in the space of the P2-P2 approximation. Therefore it's possible you are seeing some kind of "superconvergent" effect in P2-P2 which is just due to the choice of exact solution.

dewenyushu commented:
@lindsayad great fix! I tried two weeks ago to check by hand the projected nodes for a 5:3 case but did not find anything suspicious. So I am really impressed by this fix and excited to see how this will change the dual mortar convergence!

Nice to see stabilization been added, too! Has stabilization been enabled for mortar contact or is it an ongoing process?

jwpeterson commented:
@dewenyushu I don't think (this particular form of) stabilization is needed for the contact problem, at least in the formulation discussed in Popp's paper, since the constraint equation involves the Lagrange multiplier variable itself. This is in contrast to the solution continuity formulation which otherwise has a zero block on the diagonal for the constraint equations. It would be good to know for sure whether this is the case, though. I guess there are somehow no issues of LBB-stability for the dual mortar formulation?

dewenyushu commented:
> @dewenyushu I don't think (this particular form of) stabilization is needed for the contact problem, at least in the formulation discussed in Popp's paper, since the constraint equation involves the Lagrange multiplier variable itself. This is in contrast to the solution continuity formulation which otherwise has a zero block on the diagonal for the constraint equations. It would be good to know for sure whether this is the case, though. I guess there are somehow no issues of LBB-stability for the dual mortar formulation?

Thanks for the reply! Good question about the stability condition. I think the dual basis satisfies the inf-sup condition by construction. I read statements like this in this paper: 

https://www.cambridge.org/core/journals/acta-numerica/article/variationally-consistent-discretization-schemes-and-numerical-algorithms-for-contact-problems/7B66B8B7573361A299206BE15E1B325D

There should be mathematical proofs somewhere else, but I have not dig deep enough for that.

lindsayad commented:
> The latter estimate _should_ be applicable to the P2-P2 discretization case, and it predicts that you should see O(h^2) convergence in the combined norm. Therefore I was a bit surprised that you saw O(h^3) for lambda in that case, but then I remembered that you are using that quartic exact solution? I believe the true flux is quadratic in that case, so it is in the space of the P2-P2 approximation. Therefore it's possible you are seeing some kind of "superconvergent" effect in P2-P2 which is just due to the choice of exact solution.

I actually saw fourth order in lambda when using the previous MMS solution and was very puzzled before remembering exactly what you're saying. So I adjusted the MMS such that the true value of lambda is cubic on the interface. Then convergence reduced to third order.

I wonder at those a priori estimates...it seems like those estimates are always given in the context of solution continuity. I guess it's unclear to me whether they are generally applicable. 3 and 3 for the P2P2 is consistent with the 2 and 2 for P1P1.

jwpeterson commented:
> I guess it's unclear to me whether they are generally applicable.

That's a really good point actually... the gap heat conduction formulation is almost "artificial" in the sense that you introduce an extra variable to discretize the flux, but in theory you could solve the problem "directly" using only primal variables. The same isn't true for solution continuity or, in the context of some of the papers on this subject, the imposition of essential BCs using Lagrange multipliers. It is probably wrong to try and apply the same estimates in both cases. Thinking of it as a standard coupled problem, we would optimally expect O(h^{p+1}) for both variables in L2-norm, which is exactly what you observe.

jwpeterson commented:
> which is exactly what you observe.

Except for the P2-P0 case, of course. I suspect that the primal variable error is "polluted" by the error from the two-orders-lower flux discretization in that case, so that formulation is definitely suboptimal in comparison to the others for gap heat conduction.

lindsayad commented:
> Thinking of it as a standard coupled problem, we would optimally expect O(h^{p+1}) for both variables in L2-norm, which is exactly what you observe.

Yes and I also think if

> Except for the P2-P0 case, of course. I suspect that the primal variable error is "polluted" by the error from the two-orders-lower flux discretization in that case, so that formulation is definitely suboptimal in comparison to the others for gap heat conduction.

Right. If we're thinking about it as a standard coupled problem, I would intuitively expect that if we're equating the flux (proportional to the primal first derivative) to a variable representation that is not one, but two orders less than the primal, that we would be unable to capture optimal convergence.

jwpeterson commented:
BTW, you may be aware of this already, but I thought I'd just add that if you were to solve the primal-only problem and then simply differentiate the FE solution (i.e. post-process it) to obtain the boundary flux, this is known to converge as O(h^p) in L2, where p is the polynomial degree used to approximate the primal variable. Therefore, the benefit we get from the extra "work" of discretizing/solving for the flux as an extra variable is the higher-order accurate, O(h^{p+1}), approximation we get.

jwpeterson commented:
... all of which leads me to conclude that, for the gap heat conduction application, it makes the most sense to use equal-order approximations (P2-P2 or P1-P1) for the primal and flux variables. If you use e.g. P2-P1 you'll get an O(h^2) approximation to the flux... but that's the same as what you would expect to get by just differentiating/post-processing the FE solution... so there's no point in doing the extra work of defining/solving for a separate flux variable just to end up with the same accuracy. 

This argument only applies to gap heat conduction of course. I think for continuity enforcement it makes sense to use mixed-order because the error estimates are different.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/544608/) on 5bd0002 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15338/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15338/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 91e607fcf2d3e739ff0639ae77be4bf639f92e43`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/544635/) on 7870ca5 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15338/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15338/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 91e607fcf2d3e739ff0639ae77be4bf639f92e43`

lindsayad commented:
@jwpeterson do you have any thoughts on why solution continuity stabilization might be weird for second order elements? I see these weird spikes in the convergence plots for some of the stabilized cases. First order elements always look good. Moreover, second order elements without stabilization never show any spikes

![p1p0-3-to-2](https://user-images.githubusercontent.com/10248304/85348951-e3a9f980-b4b1-11ea-9cdb-b5f6eb800b98.png)
![p1p1-3-to-2](https://user-images.githubusercontent.com/10248304/85348955-e4db2680-b4b1-11ea-89d3-027ab0696853.png)
![p2p0-3-to-2](https://user-images.githubusercontent.com/10248304/85348957-e573bd00-b4b1-11ea-9043-b240718678f4.png)
![p2p1-3-to-2](https://user-images.githubusercontent.com/10248304/85348959-e60c5380-b4b1-11ea-9aae-1f5ca930d71b.png)
![p2p2-3-to-2](https://user-images.githubusercontent.com/10248304/85348961-e6a4ea00-b4b1-11ea-9c6a-9ab3a2603a5a.png)

lindsayad commented:
This is my favorite:
![p2p0-5-to-2](https://user-images.githubusercontent.com/10248304/85349009-1d7b0000-b4b2-11ea-84ba-b8905c27cddf.png)

lindsayad commented:
Need to update this branch with #15473

jwpeterson commented:
Yeah @lindsayad I don't know what would cause these jumps in the convergence plot *only* for the stabilized scheme... if it was some other issue with the creation of the mortar segments on P2 elements, you'd definitely expect it to show up in the unstabilized case as well. 

P2-P0 definitely seems to be the most erratic of the bunch, especially in your last image where the error jumps the most on the finest mesh! Maybe best to just avoid P2-P0 if we don't really understand what's going on with it, and stick to equal-order or flux-one-order-less formulations in general? Even in the P2-P0 cases that "work", the flux is only converging at 1st-order ($\ell + 1$), we see the same basic pattern in the P2-P1 case and the results are a lot smoother.

lindsayad committed:
Geometrically ghost the whole interface

lindsayad commented:
@jwpeterson the erraticism is gone with delta = 0.1:

![p2p0-5-to-2](https://user-images.githubusercontent.com/10248304/85455048-65516400-b552-11ea-9d33-2f5e3fce1d31.png)
![p2p1-5-to-2](https://user-images.githubusercontent.com/10248304/85455051-65e9fa80-b552-11ea-9d97-aceabdd39d9a.png)
![p2p2-5-to-2](https://user-images.githubusercontent.com/10248304/85455053-66829100-b552-11ea-875b-f7b725e51ad3.png)

jwpeterson commented:
> the erraticism is gone with delta = 0.1:

:+1: It's so beautiful.

jwpeterson commented:
So if I look at for example the orange and brown lines in the P2P2-5-to-2 figure, it seems they both have the same slope but the stabilized one has lower error. Similarly for the red and grey lines, although not quite as large of a gap. It seems that using stabilization is therefore both more robust (avoid occasional convergence failures) and more accurate, so win-win.

jwpeterson commented:
Oh, also worth noting is that we lose a half-power of h for the primal variable L2-norm in the P2-P0 case, i.e. we get O(h^2.5), so IMO this is similar to the gap heat conduction case where we concluded that this formulation is sub-optimal/the LM variable error pollutes the primal variable error. That's also good to know.

lindsayad commented:
Both good points. I'm satisfied that the code is beautiful now

jwpeterson commented:
OK, one last thing that seems important and I totally forgot to mention: you are actually seeing different convergence rates for the P2-P0 and P2-P2 cases depending on coarse primary/secondary. P2-P1 seems to not be affected by this choice. I will try to summarize in the tables below:

| **Coarse Primary**           |  u | LM |
|----------| --------- | -- |
| P2-P0 |      3       |  1  |
| P2-P1 |      3      |  2  |
| P2-P2 |      3       |  2  |


| **Coarse Secondary**           |  u | LM |
|----------| --------- | -- |
| P2-P0 |      *2.5* &#x1F53B;      |  1  |
| P2-P1 |      3      |  2  |
| P2-P2 |      3       |  *2.5* &#x1F53A;  |

I'm not sure whether these results are general, etc. but IMO since you generally can't guarantee whether primary or secondary will be "coarser" in real problems, P2-P1 seems like the most robust discretization since it doesn't care, P2-P0 should probably be avoided, and the extra cost of P2-P2 vs. P2-P1 isn't worth the chance of possibly getting O(h^2.5) convergence for the LM.

lindsayad commented:
Well put!

On Tue, Jun 23, 2020 at 2:14 PM John W. Peterson <notifications@github.com>
wrote:

> OK, one last thing that seems important and I totally forgot to mention:
> you are actually seeing different convergence rates for the P2-P0 and P2-P2
> cases depending on coarse primary/secondary. P2-P1 seems to not be affected
> by this choice. I will try to summarize in the tables below:
> *Coarse Primary* u LM
> P2-P0 3 1
> P2-P1 3 2
> P2-P2 3 2
> *Coarse Secondary* u LM
> P2-P0 *2.5* 🔻 1
> P2-P1 3 2
> P2-P2 3 *2.5* 🔺
>
> I'm not sure whether these results are general, etc. but IMO since you
> generally can't guarantee whether primary or secondary will be "coarser" in
> real problems, P2-P1 seems like the most robust discretization since it
> doesn't care, P2-P0 should probably be avoided, and the extra cost of P2-P2
> vs. P2-P1 isn't worth the chance of possibly getting O(h^2.5) convergence
> for the LM.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/15338#issuecomment-648431929>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOGA4H5DB5ASCV7JMWXF7TRYELNHANCNFSM4NICAFNQ>
> .
>

lindsayad committed:
Mortar distributed mesh

fdkong committed:
Reinit scaling matrix

do not delete remote elements if we already set the flag in libmesh

Refs #0000000000000000000

fdkong committed:
Sync displaced mesh and reference mesh when outputting exodus

fdkong committed:
Clone and set mesh

fdkong committed:
Set_Mesh and Clone work for ghosting functors

fdkong committed:
Added a test using distributed mesh and displaced mesh

fdkong committed:
Added a test: periodic boundary conditions + distributed mesh + displaced mesh

fdkong committed:
Added tests for distributed mesh + mesh adaptivity

fdkong committed:
Added docs for changes on mesh adaptivity

Closes #15702

fdkong committed:
Add autoscaling tests for mesh adaptivity

Closes  #15702

lindsayad commented:
Ok, resuming work on this PR. I'm going to make the correct interaction of mesh generators with distributed mesh a prereq for getting this merged, e.g. the tests in this PR must be able to to run with distributed mesh.

lindsayad commented:
Some good discussion of distributed mesh and relationship managers in #15655

lindsayad committed:
Update convergence studies tests

- Master->primary, slave->secondary
- Apparently `MeshCollectionGenerator` changed since I last worked on
  this PR, and you cannot combined meshes who have different names for
  same boundary IDs

lindsayad committed:
Add relationship managers to new MeshGenerator MeshBase objects

We do a few things:
- Split `add` and `attach` operations in `AddRelationshipManager`
- Move `add_geometric_rm` task to before mesh generation in `Moose.C`
- Create `MooseApp::attachRelationshipmanagers(MeshBase & mesh)` API
- Call said API from mesh generators that create new `MeshBase` objects,
  e.g. `FileMeshGenerator` and `GeneratedMeshGenerator`

In this fashion we can be sure that objects which add geometric
relationship managers will actually have their requests honored, even
when mesh generation is done through the `MeshGenerator` system. These
changes also make it possible for `MeshGenerators` themselves to
request geometric ghosting from upstream `MeshGenerators`

lindsayad committed:
Generalize AugmentSparsityOnInterface

`AugmentSparsityOnInterface` should now work when called during
intermediate stages of mesh generation, e.g. if a `MeshGenerator` calls
`prepare_for_use` before the whole mesh generation process is complete.
This works by assuming the worst: e.g. if we are still generating the
mesh, then we don't know if our mortar boundary ids or subdomain ids
exist yet, so it's safest to ghost any higher dimensional elem that lies
on a boundary or any lower dimensional elem whose interior parent lies
on the boundary

lindsayad committed:
Don't reattach geometric ghosting functors

lindsayad committed:
Add MooseMesh::allowRemoteElementRemoval

This API makes it easy to communicate not to remove element removal when
the problem has a displaced mesh and the mesh is being generated using
mesh generators

lindsayad committed:
One missing rm clone

lindsayad committed:
Make LowerDBlockFromSidesetGenerator agnostic of mesh parallelism

lindsayad committed:
Sanity check in ProxyRelationshipManager

lindsayad committed:
Demonstrate working mortar on displaced and distributed mesh

lindsayad committed:
Add RelationshipManager::init(MeshBase &) method

lindsayad committed:
Call meshChanged after ghostGhostedBoundaries

lindsayad committed:
Account for the fact that ghosting functors can be cloned

`MeshExtruderGenerator` for example clones the source mesh when creating
the destination mesh, which will clone the ghosting functors. Hence
straight pointer comparison is not valid. Instead we will call
RelationshipManager::operator== to determine whether we've retained all
the geometric ghosting functors/relationship managers that we added
during the mesh generation phase

lindsayad committed:
Late geometric ghosting should always turn off early elem deletion

This commit also refactors the relationship manager init process into
something that's hopefully a little more simple and robust. We also
simplify RM addition such that `Action` should leave all the decision
making to `MooseApp`. This is now passing three out of four of Fande's
dmg bc tests

lindsayad committed:
Encapsulate with MooseMesh::deleteRemoteElements

lindsayad committed:
Let ghosting functors handle ghosting for periodic

lindsayad committed:
post_equation_systems_init -> after_late_geometric_ghosting

lindsayad committed:
Simplify AugmentSparsityOnInterface

Also clean up some comments

lindsayad committed:
Dont serialize with exodus when doing CheckGhostedBoundaries

lindsayad commented:
@fdkong do you want to merge #15655 into this? I currently have all your commits from there on here. At this point I don't see much of a reason to divorce the mortar stuff from other code in this PR because the mortar convergence study portion of the PR should be relatively small in terms of % line-count. I could try though to split out the commits that are meant to help #15655 but I'm not sure if I would be able to.

If we do decide to combine, I know this PR needs gold nemesis files for your new test(s)

fdkong commented:
> If we do decide to combine, I know this PR needs gold nemesis files for your new test(s)

Yes, we could roll together

lindsayad committed:
Add missing nemesis files for Fande's test

lindsayad committed:
Account for elem deletion in unique id numbering

lindsayad committed:
Move RM attachment and setting elem removal flg to buildMeshBaseObject

This makes it so that anyone adding a new `MeshGenerator` that builds a
`MeshBase` object doesn't have to know that they should add relationship
managers and toggle element deletion flags, so it's less prone to error.
Also we should make all `MeshGenerators` go through
`buildMeshBaseObject`.

jessecarterMOOSE commented:
Hey @lindsayad @fdkong - I checked out this PR in order to do mesh adaptivity and automatic scaling, which does work, but when I add a few simple materials and then try to do a LineMaterialRealSampler VectorPostprocessor, I get a segfault, even if I don't use scaling. For example, adding the following to [cycles_per_step.i](https://github.com/idaholab/moose/blob/next/test/tests/adaptivity/cycles_per_step/cycles_per_step.i) is enough to cause the segfault.

```
[Materials]
  [./const]
    type = GenericConstantMaterial
    prop_names = 'A B'
    prop_values = '1.0 2.0'
  [../]
[]

[VectorPostprocessors]
  [./mats]
    type = LineMaterialRealSampler
    start = '0 0 0'
    end = '10 10 0'
    property = 'A B'
    sort_by = 'x'
  [../]
[]
```

After a few timesteps it spits out this:
```

No index 18446744073709551615 in ghosted vector.
Vector contains [0,2960)
And empty ghost array.
```

And then a bunch of debugging information that you would probably find useful but I want to clog up the chat.

lindsayad commented:
> when I add a few simple materials and then try to do a LineMaterialRealSampler VectorPostprocessor, I get a segfault, even if I don't use scaling.

Fun! This still has some failing framework tests to resolve. Once I resolve those, I will look test your input and see what's going on. Appreciate the bug report!

lindsayad committed:
attachRelationshipManagers now takes a MooseMesh arg

The GridPartitioner actually creates its own mesh, which renders
comparisons between the MeshBase of the MooseMesh from the action
warehouse and the incoming MeshBase invalid. Thus we should just pass
MooseMesh as an argument so the checks are always guaranteed to work.

lindsayad committed:
Tighten tolerance and regold dmg_periodic_bc_out_3d.e

lindsayad committed:
Num touchable periodic nodes not reliable in distributed mode now

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit b51a1fac743ce11a1e65018f25a478241c24ed44
+Subproject commit 16f1801381b3a63b644002330b3d029eed35ff3b
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

lindsayad commented:
This periodic simulation is really pissing me off

fdkong commented:
> This periodic simulation is really pissing me off

Sorry, buddy

IIRC, the source is discontinuous, and there is only a small region with nonzero values.  That somehow makes the solution in the transition zone different in different runs. The original loose tolerances I added do not work? You do not need to tighten the tolerances.

If it  is only the test that blocks you, you could skip it (or you use ridiculous tols as well) for a moment. I will investigate into it once I have a chance. As you know, I have to work on another project at this point.

lindsayad committed:
Make 3d test very simple

With the new 3d test the value on the block should be equal to the
current timestep, and this is what I see with 2 and 3 procs. However,
running with 4 procs, the result is very wrong.

lindsayad committed:
Let GhostPointNeighbors handle geometric periodic bc ghosting

lindsayad committed:
Don't remove default ghosting from the mesh for DisplacedProblem

For quite a while now we haven't removed default ghosting from the mesh
from `FEProblemBase`. If we have `ProxyRelationshipManagers` for
instance, we rely on having identical ghosting from the reference mesh
to the displaced mesh (indeed we should have this no matter what). If
default ghosting is missing from the `DisplacedProblem` but not from the
`FEProblemBase` then there is a
very strong chance that ghosting will not be identical between the two.
This is certainly true now that I've added periodic boundary ghosting to
the default libMesh `GhostPointNeighbors` ghosting functor.

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit b51a1fac743ce11a1e65018f25a478241c24ed44
+Subproject commit abb231493bea946ae55e9b39828ff9fd55daa60f
```
**Caution!** This contains a submodule update

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit b51a1fac743ce11a1e65018f25a478241c24ed44
+Subproject commit 3f9015493934a777ff9228616e613144a664de7c
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

lindsayad committed:
Turn off recover for distributed + pbcs + adaptivity + nemesis

lindsayad committed:
Add back RelationshipManager::internalInit()

This is for backwards compatability with user code that implemented the
empty `internalInit()` method. This also requires a rename of
`internalinit(MeshBase &) -> internalInitWithMesh(MeshBase &)` in order
to avoid warnings about `-Woverloaded-virtual`

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit b51a1fac743ce11a1e65018f25a478241c24ed44
+Subproject commit ef2cacfd227f75f812f163dad2e4c1b61e1e75e1
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

lindsayad committed:
Restrict couple dmg tests to non-debug

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit b51a1fac743ce11a1e65018f25a478241c24ed44
+Subproject commit f33331f6ebb0f31dff7526106268ca0c4300049e
```
**Caution!** This contains a submodule update

moosebuild commented:
Job [Python 3.7 Tests](https://civet.inl.gov/job/597486/) on 11a7ffb : invalidated by @lindsayad

moosebuild commented:
Job [Python 3.6 Tests](https://civet.inl.gov/job/597485/) on 11a7ffb : invalidated by @lindsayad

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/597496/) on 11a7ffb : invalidated by @lindsayad

moosebuild commented:
Job [Modules debug PETSc alt](https://civet.inl.gov/job/597488/) on 11a7ffb : invalidated by @lindsayad

lindsayad commented:
Oh wow such beautiful green. Just dying to be reviewed

jwpeterson commented:
This PR has been quite a journey!

fdkong commented:
> Oh wow such beautiful green. Just dying to be reviewed

So beautiful, I should merge it instead of review it

fdkong commented on framework/src/actions/AddRelationshipManager.C:
```diff
@@ -14,6 +14,9 @@
 registerMooseAction("MooseApp", AddRelationshipManager, "attach_coupling_rm");
+registerMooseAction("MooseApp", AddRelationshipManager, "add_geometric_rm");
+registerMooseAction("MooseApp", AddRelationshipManager, "add_algebraic_rm");
+registerMooseAction("MooseApp", AddRelationshipManager, "add_coupling_rm");
```
What are differences between "attach" and "add"?

lindsayad replied:
If you have mesh generators then RMs are "attached" to the `MeshBase`  object(s) at the same time that they are added, so no difference. When we do not have mesh generators, then attaching happens during a different phase than adding. During adding, we store the RMs in a container on `MooseApp`, and then during attaching we actually attach them to the mesh or dof-map. I *think* this was done because sometimes you cannot attach geometric relationship managers early in the simulation...like a geometric rm associated with periodic boundaries or something

fdkong commented on framework/src/base/Assembly.C:
```diff
@@ -2654,25 +2681,25 @@ Assembly::prepareLowerD()
 
       // derivatives w.r.t. interior secondary residuals
-      jacobianBlockLower(Moose::SecondaryLower, vi, vj, tag)
+      jacobianBlockMortar(Moose::SecondaryLower, vi, vj, tag)
```
Motivation that you changed "Lower" to "Mortar"?

lindsayad replied:
I think I had a good reason at the time, but I've forgotten it now :cry:

lindsayad replied:
Ah it's because you can retrieve Jacobian blocks from this method that are totally unrelated to the lower-dimensional element degrees of freedom, e.g you can retrieve `Moose::PrimaryPrimary, Moose::PrimarySecondary` etc. So I thought that `jacobianBlockLower` was a bad name

fdkong commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -44,24 +46,29 @@ AutomaticMortarGeneration::AutomaticMortarGeneration(
+  if (_distributed)
+    mortar_segment_mesh = libmesh_make_unique<DistributedMesh>(mesh.comm());
+  else
+    mortar_segment_mesh = libmesh_make_unique<ReplicatedMesh>(mesh.comm());
```
`_distributed` should simply derive from the parent  mesh. For example, if I am using 3D distributed mesh, I would expect 2D mortar mesh is distributed as well.

lindsayad replied:
That's why I'm doing...right?

fdkong commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -119,6 +126,25 @@ AutomaticMortarGeneration::buildMortarSegmentMesh()
+    const auto num_secondary_nodes =
+        std::distance(mesh.bid_nodes_begin(secondary_bnd_id), mesh.bid_nodes_end(secondary_bnd_id));
+
+    node_unique_id_offset += num_primary_nodes + 2 * num_secondary_nodes;
```
It would be great if you could have documents how you come up these magic offsets.

lindsayad replied:
I have this comment ten lines earlier:
```
  // Create an offset by the maximum number of mortar segment elements that can be created *plus*
  // the number of lower-dimensional secondary subdomain elements. Recall that the number of mortar
  // segments created is a function of node projection, *and* that if we split elems we will delete
  // that elem which has already taken a unique id
```

fdkong commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -133,22 +159,30 @@ AutomaticMortarGeneration::buildMortarSegmentMesh()
+      new_elem = new Edge3;
     else
-      new_elem = mortar_segment_mesh.add_elem(new Edge2);
+      new_elem = new Edge2;
```
Mortar mesh is always 1D?

lindsayad replied:
Yea we do not do mortar with 3D (e.g. 2D mortar segment mesh elements) yet

fdkong commented on framework/src/relationshipmanagers/AugmentSparsityOnInterface.C:
```diff
@@ -78,79 +91,132 @@ AugmentSparsityOnInterface::operator()(const MeshBase::const_element_iterator &
+        mooseAssert(secondary_subdomain_id != Moose::INVALID_BLOCK_ID,
+                    "Secondary subdomain id should exist by now. If you're using a MeshModifier "
+                    "please use the corresponding MeshGenerator instead");
+
```
Roy styles, Assert everywhere , haha

lindsayad replied:
It helps me keep track of what I'm doing and I grow more and more paranoid over the years :smile:

fdkong commented on test/tests/bcs/dmg_periodic/tests:
```diff
@@ -12,12 +12,9 @@
 
   [./check_one_step]
-    type = 'Exodiff'
+    type = 'RunApp'
```
Still have diff issues?

lindsayad replied:
I don't think so but maybe I didn't change it back to `Exodiff` because it seems like the purpose of this test (based on the requirement) is just to make sure that the boundaries are actually distributed? And there are a lot of other `Exodiff` tests in this directory

fdkong replied:
Yes

fdkong commented on test/tests/mesh/periodic_node_map/tests:
```diff
@@ -8,6 +8,7 @@
 
     requirement = "The system shall produce a valid periodic node map with only one periodic constraint direction."
     recover = false  # Steady Test (Needs to be here since we don't want the checkpoints clobbering each other)
+    mesh_mode = 'replicated' # We no longer ghost all periodic boundaries for all procs
```
mesh_mode = 'replicated distributed'? I want to test both

lindsayad replied:
These tests use a custom object that expects there to be a certain number of entries in their periodic node map. We no longer ghost all boundaries just for grins, so we no longer have all the boundary nodes when we are running distributed

fdkong replied:
Nice!

fdkong reviewed (changes requested):
It looks great for me.

Should we ask for one more judge  because I have a few commits in this PR?

lindsayad commented on framework/include/base/Assembly.h:
```diff
@@ -389,6 +395,22 @@ class Assembly
+  /*
+   * @return The current lower-dimensional element volume
+   */
+  const Real & lowerDElemVolume() const { return _current_lower_d_elem_volume; }
```
This is for mortar calculations. The lower dimensional elements have to be explicitly added by the user during mesh generation.

lindsayad commented on framework/include/base/Assembly.h:
```diff
@@ -834,11 +861,11 @@ class Assembly
-   * like MortarConstraints
+   * Add *all* portions of the Jacobian, e.g. LowerLower, LowerSecondary, LowerPrimary,
+   * SecondaryLower, SecondarySecondary, SecondaryPrimary, PrimaryLower, PrimarySecondary,
+   * PrimaryPrimary for mortar-like objects
```
Correct, this is for mortar generally. We came up with primary-secondary terminology as a replacement for master-slave. I'll see about adding a bit more to this doxygen comment

lindsayad commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -267,4 +265,7 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
   const bool _periodic;
+
+  /// Whether the mortar segment mesh is distributed
+  const bool _distributed;
```
I do.
```c++
AutomaticMortarGeneration::AutomaticMortarGeneration(
    MooseApp & app,
    MeshBase & mesh_in,
    const std::pair<BoundaryID, BoundaryID> & boundary_key,
    const std::pair<SubdomainID, SubdomainID> & subdomain_key,
    bool on_displaced,
    bool periodic)
  : ConsoleStreamInterface(app),
    mesh(mesh_in),
    h_max(0.),
    _debug(false),
    _on_displaced(on_displaced),
    _periodic(periodic),
    _distributed(!mesh.is_replicated())
{
```

lindsayad commented on framework/include/constraints/MortarConstraintBase.h:
```diff
@@ -152,20 +151,16 @@ class MortarConstraintBase : public Constraint,
+  Elem const * const & _lower_secondary_elem;
+
+  /// The primary face lower dimensional element
+  Elem const * const & _lower_primary_elem;
```
It is not the mortar segment element. There are the lower-dimensional secondary and primary elements that correspond to faces of higher dimensional elements on the mortar boundary. The  mortar segment elements are formed starting with the lower-dimensional secondary element, and the projecting nodes from the lower-dimensional primary element and splitting those elements where the nodes project

lindsayad replied:
I will add more doxygen for this

lindsayad committed:
More doxygen

lindsayad commented:
@fdkong I added a little bit more doxygen. I suppose we can ask @aeslaughter , @loganharbour , or @rwcarlsen for an additional review?

lindsayad commented:
If we don't get another review, I'll merge this at the end of Monday.

lindsayad commented:
Ruh roh, looks like periodic distributed recover without all ghosted boundaries has some problems https://civet.inl.gov/job/603572/

lindsayad commented:
@roystgnr I should know this but does checkpoint write out ghost elements? I would assume so

fdkong commented:
> @roystgnr I should know this but does checkpoint write out ghost elements? I would assume so. Yea I see some `query_ghosting_functors` in Checkpoint IO.

Data is "distributed" in the file. When we read it back, we may not connect the whole story together.  So we need communicate data around ... to figure out the whole  story

roystgnr commented:
> I should know this but does checkpoint write out ghost elements? I would assume so. Yea I see some query_ghosting_functors in Checkpoint IO.

It does, but I'm afraid there's a bit of a trick to it: we write out ghost elements but *not* ghosting functors, much less the data that particular functors depend on.  I guess we could make that possible, albeit at great difficulty, if we added a factory pattern and also required a serialize routine in the base class to be overridden?  But at the moment the way it works is we expect you to have the same ghosting requirements set up for the read() that you did for the write() ... so if you haven't yet set up all your periodic BCs, for example, then we'll happily read the periodic ghosted elements, discover you haven't said they're still necessary, and discard them.

In the future, this could be one of the use cases of that post-prepare_for_use idiom we're brainstorming.  A CheckpointIO::read() would leave need_remote_elem_removal() (or whatever we call the status accessor for that particular substep) set to false, so subsequent preparation work would skip remote element deletion and you'd have more time to get your ghosting functors and PBCs and algebraic coupling functors (this bit @arovinelli just last week!) and everything into place first.  But unlike allow_remote_element_removal(false), the new setting wouldn't be *permanent*, so if you do AMR or load balancing or reduce ghosting extent or anything like that later in the simulation, those operations would naturally set need_remote_elem_removal(true) and you wouldn't have to worry about whether you might still have no-longer-necessary ghosted elements floating about afterward.  But the whole process would be more robust to different user idioms: instead of having to get your ghosting functors set up before the restart's first prepare_for_use() you'd merely have to get them set up before they're first truly needed.
